### PR TITLE
add a function to process decimal type comparing with floating point literal  in IN operator

### PR DESCRIPTION
**Pull request by lhuang09287750**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
add a support for decimal type comparing with floating point literal  in IN operator

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Interpreters/convertFieldToType.cpp [hunk 1]
```diff
@@ -109,6 +109,20 @@ Field convertDecimalToDecimalType(const Field & from, const DataTypeDecimal<T> &
     return DecimalField<T>(value, type.getScale());
 }
 
+template <typename From, typename T>
+Field convertFloatToDecimalType(const Field & from, const DataTypeDecimal<T> & type)
```
Comment (vdimir) [inline src/Interpreters/convertFieldToType.cpp:113]: Consider `DecimalUtils::convertTo<Float64>(...)`

Comment (lhuang09287750) [reply src/Interpreters/convertFieldToType.cpp:113]: thank you for advice.
I've used **convertToDecimal** in my func **convertFloatToDecimalType**
that works well

Comment (vdimir) [reply src/Interpreters/convertFieldToType.cpp:113]: @lhuang09287750 Sorry for the delay

The main concern is converting float to string, which looks like a workaround.

What about using a scale of the target type and then checking that we didn't lose any precision (as I did for decimals with different scales)

Here are the changes, you can cherry-pick commits or force-push my branch (https://github.com/ClickHouse/ClickHouse/tree/vdimir/decimal-in-float) into the current PR: https://github.com/ClickHouse/ClickHouse/compare/vdimir/decimal-in-float

Comment (lhuang09287750) [reply src/Interpreters/convertFieldToType.cpp:113]: back from vacation
I've cherry-picked your commits and made a push again!
Your fix also worked,but I don't understand why, what does the function in ActionVistors.cpp do?

Comment (lhuang09287750) [reply src/Interpreters/convertFieldToType.cpp:113]: I make a debug and understand it! :->

Comment (vdimir) [reply src/Interpreters/convertFieldToType.cpp:113]: > what does the function in ActionVistors.cpp do?

It converts decimal back to float and check that it's the same number, so no precision loss.

**Review by vdimir — APPROVED** [review]
Thank you for your contribution!

I'll rebase the branch (some build was failed, but it's not related to changes) and merge if CI is ok.

**Pull request merged**